feat(contrib): uart_reader Version 4 submission

vctrcc opened the pull request:
I have cleaned up a bit the repository by deleting old scripts, adjusting rye, and a few other touches.

![image](https://github.com/user-attachments/assets/2d8d1b61-6a51-4ddb-af5c-d11f149ce994)
--
![image](https://github.com/user-attachments/assets/5a88c32b-f781-4512-a70e-e9c34303a418)

Thank you for the feedback.

vctrcc commented:
It seems that some files from one of the remote repo's commits got added to the pull request for some reason.

jeertmans commented on mcu/src/mcu/uart_reader.py:
```diff
@@ -31,17 +31,11 @@
-    import mcu.libraries.loggingUtils as logu
-    import mcu.libraries.serialUtils as seru
-    from mcu.model_formatter_for_V4 import *
+# Custom modules (if not running, remove the mcu. prefix)
+import mcu.libraries.database_utils as dbu
+import mcu.libraries.logging_utils as logu
+import mcu.libraries.serial_utils as seru
+from mcu.model_formating_template import *
```
Small suggestion, doesn't relative imports work here?
```suggestion
# Custom modules (if not running, remove the mcu. prefix)
import .libraries.database_utils as dbu
import .libraries.logging_utils as logu
import .libraries.serial_utils as seru
from .model_formating_template import *
```

Note that start imports, i.e., `from X import *`, are usually bad practice.

vctrcc replied:
From what i have seen, only this method works :
from .libraries import database_utils as dbu

Is this still better than just doing mcu.libraries.x ?

jeertmans replied:
> I will try it again, but last time the compiler complained (ill also change the star import).

The compiler?

jeertmans replied:
> From what i have seen, only this method works : from .libraries import database_utils as dbu
> 
> Is this still better than just doing mcu.libraries.x ?

I'll edit your pull request, it's easier for me :-)

vctrcc replied:
>The compiler?

I meant the interpreter

vctrcc commented:
Little update, i have found a quite inconvenient bug that i could only find a workaround with, its the bug with the model creation, for some reason, the pickle cannot be resolved if the model was not created using rye. So rye is a necessity.

Also, here is the current progress of the repair, please advise if you see more things to add : 
<img width="281" alt="image" src="https://github.com/user-attachments/assets/6186a386-e56c-4a8d-8824-d629834f5e4b" />
The Last 3 need checking if they where correctly fixed, as my board is currently non operational (fast blinking red LED3), and the rest of the group require the other board for this week's work.

jeertmans commented:
See my last two commits @vctrcc, you should in theory just clean [contrib/README.md](https://github.com/LELEC210X/LELEC210X/pull/54/files#diff-549baf4b2300e19b6400b465ecd39ef75bc87d7b6ef6b6eb8800c5658f9c4b89) so it contains a good documentation. I recommend to avoid including all the CLI options, but rather mention you can get them with `rye run uart-reader --help`.

**Important:** I have renamed to entrypoint from `uart_reader` to `uart-reader`.

vctrcc commented:
Small problem, all files have been deleted, and now there is no more uart reader anywhere :/

vctrcc commented:
I will not modify much more the code, appart if i find a bug (but since i don't have a working stm32 board, i can't test anything at least till friday). I will start writing the README tonight. If you can start reviewing stuff related to the code it would be awesome, but i understand if its best to wait a bit. 

Could you be the one making the tag for the old version ? I don't understand exactly what i have to do, and i don't think i have access to making tags in the main repo.  (it would be usefull to be able to add the tag in the README)

In short : The code should be ready, but i have yet to re-test everything with a stm32 nucleo board. I will write the README tonight.

jeertmans commented:
No worries, just press the ready for review button when you are :)

Should I create the tag from the version that is currently on our main branch?

vctrcc commented:
Yes, if you can, and ill tag it as ready for merge once i have a bit more written on the readme

jeertmans commented on contrib/README.md:
```diff
@@ -0,0 +1,324 @@
+If you think you also have a piece of code that is worth adding to this project,
+please do not hesitate to reach out (via a pull request or by contacting a teaching assistant)!
+
+## List of contributions
+
+- [Contribution 1 - UART Reader](#uart-reader)
+  - [UART Reader Manual](#uart-reader-in-depth-manual)
```
```suggestion
**List of contributions:**

- [UART Reader](#uart-reader)
  - [Manual](#uart-reader-in-depth-manual)
```

I'd avoid the extra wording and the unnecessary heading.

The documentation seems very long. I'd keep it short, and maybe create a second markdown file, `UART_READER.md`, that contains more details.

jeertmans commented on contrib/README_old.md:
To be deleted

jeertmans commented on contrib/assets/app_architecture.pdf:
Only keep one of the two: PNG or PDF, not both.

jeertmans commented on contrib/assets/databaseV2_picture_init.PNG:
Avoid *V2* inside the name of the file.

Probably also rename all image files by prepending `uart_reader_`. For example, `uart_reader_app_arch.png`. This will be easier to add other contributions and not mess with files.

jeertmans commented on contrib/assets/uart_reader_GUI_V1.png:
Same comment about *V1*

vctrcc replied:
Then how should i refer to the V1 and V4 of the gui ? old and new ?

jeertmans commented on contrib/src/contrib/uart_reader/__main__.py:
```diff
@@ -0,0 +1,1700 @@
+    db.add_item(
+        "_hidden",
+        "appversion",
+        db.ConstantText("App Version", "4.1", "The version of the application"),
```
```suggestion
        db.ConstantText("App Version", __version__, "The version of the application"),
```

This is probably better to create a `__version__.py` file in `src/contrib/uart_reader.py` that has a version string `__version__ = "4.1"` inside, and you import the version at the top of `__main__.py`: `from .__version__ import __version__`.

vctrcc commented:
I have 3 shemas for the app architecture, i can't decide witch one is the best, could you advise me ?

vctrcc commented:
The readme is about 80% donne, ill continue to write more
After today's demo, and fixing the problem with my nucleo board, ill test once again everything
Then the only thing left will be making the git tag

vctrcc commented:
@jeertmans I have finished everything, and tested everything i could. Its all ready to be reviewed.
(i just remembered that i have to update a image, but that will be the last commit)

jeertmans reviewed (changes requested):
Looks good, just clean and rename the images :)
Ping me when it is done

vctrcc commented:
@jeertmans I have removed the pictures i was not using, removed the v1 and v4 in the picture names, and changed the README's to point to the right images and references.
--> Everything should be clean

jeertmans commented:
Looks great to me, thanks for your contribution :)

vctrcc commented:
Thank you for your help !
